python_modules/dagster-graphql/dagster_graphql/implementation/loader.py

```diff
@@ -317,20 +328,37 @@ def _build_cross_repo_deps(
                     depended_by=[],
                 )
 
-        return sink_assets, external_asset_deps
+        return sink_assets, external_asset_deps, all_deps
 
     def get_sink_asset(self, asset_key: AssetKey) -> ExternalAssetNode:
-        sink_assets, _ = self._build_cross_repo_deps()
+        sink_assets, _, _ = self._build_dep_data()
         return sink_assets[asset_key]
 
     def get_cross_repo_dependent_assets(
         self, repository_location_name: str, repository_name: str, asset_key: AssetKey
     ) -> Sequence[ExternalAssetDependedBy]:
-        _, external_asset_deps = self._build_cross_repo_deps()
+        _, external_asset_deps, _ = self._build_dep_data()
         return external_asset_deps.get((repository_location_name, repository_name), {}).get(
             asset_key, []
         )
 
+    def get_cycle_for_key(self, key: AssetKey) -> Optional[Sequence[AssetKey]]:
+        return self._get_key_to_cycle_map().get(key)
+
+    def has_cycles(self) -> bool:
+        return len(self._get_key_to_cycle_map()) > 0
+
+    @lru_cache(maxsize=1)
```

Comment:
wonder should use our `cached_method` instead to more clearly scope the one copy to the object . Not clear to me how `@lru_cache(maxsize=1)` would behave in multitenant web server context where N of these objects can exist

Reply:
IIUC `lru_cache` and `cached_method` behave identically with respect to their logic (i.e. both maintain separation of caching per-object). The difference is that the cache on `cached_method` is an instance attribute and the cache on `lru_cache` is a class attribute (the instance is used as a parameter in calculating the cache key). Using `cached_method` therefore reduces memory burden with more granular garbage collection.

In any case the reason I used `lru_cache` is because there was already a method on this loader using it, but I can convert them both if you think that's a good idea-- I can't think of why we _wouldn't_ do it (more GC seems better) but I'm hardly an expert here.

Reply:
> The difference is that the cache on cached_method is an instance attribute and the cache on lru_cache is a class attribute (the instance is used as a parameter in calculating the cache key)

ya the concern I had is that with class level `maxsize=1` if there are multiple concurrent asset graph resolutions happening that they could fight over the one slot and result in recomputing repeatedly 

> convert them both if you think that's a good idea
>  I can't think of why we wouldn't do it

ya i think it makes sense to switch them